### DISCUSSION TITLE
Livewire 3 - How to listen for change event in Javascript

### DISCUSSION THREAD

**Discussion by ZebTheWizard** [livewire/livewire#6295] [1. Help]
I have a Livewire component that saves a model to the database and then flashes the session with a message.  However, I want to replace the session message with an Alpine toast component.  How can I invoke javascript inside an Alpine component from my Livewire 3 component? Trying to do something like this:

```php

class CreatePost extends Component
{

    public function save()
    {
        Post::firstOrCreate([
            'title' => $this->title,
            'content' => $this->content,
        ]);

        $this->js(<<<JAVASCRIPT
              title = "successfully submitted content"
              popToast();
        JAVASCRIPT);
    }
}
```

**Comment by N1ebieski** [comment]
Use events > https://livewire.laravel.com/docs/events#events-in-alpine

**Comment by proofmancer** [comment]
To invoke JavaScript within an Alpine component from your Livewire component, you can utilize Livewire's built-in JavaScript methods. You can use the dispatchBrowserEvent method to send a browser event to trigger the desired JavaScript action in your Alpine component. Here's how you can achieve this:

First, make sure you're emitting a browser event from your Livewire component after saving the model and before setting the JavaScript that triggers the toast in your Alpine component.

```
class CreatePost extends Component
{
    public $title;
    public $content;

    public function save()
    {
        Post::firstOrCreate([
            'title' => $this->title,
            'content' => $this->content,
        ]);

        $this->dispatchBrowserEvent('show-toast', ['message' => 'successfully submitted content']);
    }

    public function render()
    {
        return view('livewire.create-post');
    }
}

```

In your Blade view (livewire/create-post.blade.php), you can set up your Livewire component and include your Alpine toast component. Be sure to listen for the browser event and use it to trigger the Alpine toast component.

```
<div x-data="{ show: false, message: '' }" x-init="
    @this.on('show-toast', data => {
        show = true;
        message = data.message;
    });
">
    <button wire:click="save">Save</button>
    
    <!-- Include your Alpine toast component here -->
    <div x-show="show">
        <div x-text="message" x-transition:enter="transition ease-out duration-300" x-transition:enter-start="opacity-0 transform scale-90" x-transition:enter-end="opacity-100 transform scale-100" class="bg-white border rounded-md shadow-sm p-4 fixed bottom-4 right-4"></div>
    </div>
</div>

```

In this example, the x-init directive listens for the show-toast browser event and then triggers the display of the Alpine toast component by updating the show and message properties.

Remember that you need to ensure you've set up Livewire and Alpine properly in your application. Additionally, the actual markup and styles of the Alpine toast component depend on how you've designed it in your application. Adjust the code as needed to match your specific setup.

**Reply by christoferd** [reply]
Does the browser event happen before or after the component is re-rendered?

**Reply by whousehomestay** [reply]
dispatchBrowserEvent has been changed to dispatch in v3.
